src/metabase/public_settings/premium_features.clj

```diff
@@ -240,3 +241,169 @@
+  (let [{:keys [feature fallback]} options]
+    `(if ~(or (= feature :none)
+              (if (or (not feature) (= feature :any))
+                `(enable-enhancements?)
+                `(has-feature? ~feature)))
+
+       ~(if result-schema
+          `(s/validate ~result-schema (do ~@body))
+          `(do ~@body))
+
+       ~(cond
+          (= fallback :error)
+          `(throw (missing-premium-token-exception ~(str fn-name) ~feature))
+
+          (or (symbol? fallback) (seq? fallback))
+          `(apply ~fallback ~args)
+
+          ;; :oss and default case
+          :else
+          `(apply (get @ee-registry [(symbol ~(str (ns-name *ns*)))
+                                     (symbol ~(str fn-name))])
+                  ~args)))))
+
+(def resolve-ee
+  "Tries to require an enterprise namespace and resolve the provided function. Returns `nil` if EE code is not
+  available, the function is not found, or any other error occurs. Memoized to avoid unecessary repeat calls to
+  `classloader/require` and `ns-resolve`."
+  (memoize
+   (fn [ee-ns fn-name]
+     (u/ignore-exceptions
+      (classloader/require (symbol ee-ns))
+      (ns-resolve (symbol ee-ns) (symbol fn-name))))))
+
+(defmacro defenterprise-oss
+  "Impl macro for `defenterprise` when used in an OSS namespace. Don't use this directly."
+  [{:keys [fn-name ee-ns docstr args body result-schema]}]
+  (register-mapping! fn-name ee-ns args body)
+  `(if-let [ee-fn# (resolve-ee ~(str ee-ns) ~(str fn-name))]
+     (apply ee-fn# ~args)
+     (let [result# (do ~@body)]
+       (if ~result-schema
+         (s/validate ~result-schema result#)
+         result#))))
+
+(defmacro defenterprise
+  "Defines a function that has separate implementations between the Metabase Community Edition (CE, aka OSS) and
+  Enterprise Edition (EE).
+
+  When used in a CE namespace, defines a function that should have a corresponding implementation in an EE namespace
+  (using the same macro). The EE implementation will be used preferentially to the CE implementation if it is available.
+  The first argument after the function name should be a symbol of the namespace containing the EE implementation. The
+  corresponding EE function must have the same name as the CE function.
+
+  When used in an EE namespace, the namespace of the corresponding CE implementation does not need to be included --
+  it will be inferred automatically, as long as a corresponding [[defenterprise]] call exists in an CE namespace.
+
+  Two additional options can be defined, when using this macro in an EE namespace. These options should be defined
+  immediately before the args list of the function:
+
+  ###### `:feature`
+
+  A keyword representing a premium feature which must be present for the EE implementation to be used. Use `:any` to
+  require a valid premium token with at least one feature, but no specific feature. Use `:none` to always run the
+  EE implementation if available, regardless of token. (Default: `:any`)
```

Comment:
I think it might make sense for `:feature` to be required since every bit of EE is supposed to be behind a feature flag and leaving it out is almost certainly a mistake. If you want to do `:any` then you should have to manually specify that I think -- it makes it harder to use this code wrong

Reply:
There are actually still a lot of places where we currently just do an `enable-enhancements?` check but don't require a specific feature -- for example, LDAP, and in the enterprise version of scoring for search results. I'll open a tech debt issue to resolve these cases.

In the mean time I'll make `:feature` required but keep the `:any` case, if explicitly desired.

Reply:
Yeah, all of that stuff **should** be using a specific feature instead. `enable-enhancements?` was sort of a placeholder because we didn't have appropriate features and we never got around to updating some of that stuff. The LDAP stuff should be using `:sso` for example. Not sure about search. But yeah we should go in and fix that stuff up. 

That's my main reason for saying I think feature should be required -- as we move stuff over to `defenterprise` it will force us to fix that stuff